src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+            EditorCompletion.CompletionItem uniqueItem = null;
+            int selectedItemIndex = 0;
+
+            // TODO: Can we get away with less complexity here by only doing hard select on preselection and not on regular filter text matching / etc...
+            // https://github.com/dotnet/roslyn/issues/29108
+
+            if (bestItem != null)
+            {
+                selectedItemIndex = itemsInList.IndexOf(i => Equals(GetOrCreateRoslynItem(i.CompletionItem), bestItem));
+                if (selectedItemIndex > -1 && bestItem != null && matchingItems.Length == 1 && filterText.Length > 0)
+                {
+                    uniqueItem = highlightedList[selectedItemIndex].CompletionItem;
+                }
+            }
+
+            var updateSelectionHint = GetUpdateSelectionHint(bestItem, snapshot, caretPosition, triggerReason, filterText, selectedItemIndex, mustSetSelection);
+
+            if (selectedItemIndex == -1)
+            {
+                selectedItemIndex = 0;
+            }
+
+            return new EditorCompletion.FilteredCompletionModel(highlightedList, selectedItemIndex, updatedFilters, updateSelectionHint, centerSelection: true, uniqueItem);
+        }
+
+        private static RoslynCompletionItem GetOrCreateRoslynItem(EditorCompletion.CompletionItem item)
+        {
+            if (item.Properties.TryGetProperty<RoslynCompletionItem>(CompletionItemSource.RoslynItem, out var roslynItem))
+            {
+                return roslynItem;
+            }
+
+            return RoslynCompletionItem.Create(item.DisplayText, item.FilterText, item.SortText);
+        }
+
+        private EditorCompletion.UpdateSelectionHint GetUpdateSelectionHint(
+            RoslynCompletionItem bestItem, ITextSnapshot snapshot, int? caretPosition, 
+            EditorCompletion.CompletionTriggerReason triggerReason, string filterText, int selectedItemIndex, bool mustSetSelection)
+        {
+            if (bestItem == null || mustSetSelection)
+            {
+                return EditorCompletion.UpdateSelectionHint.SoftSelected;
+            }
+
+            // We don't have a builder and we have a best match.  Normally this will be hard
+            // selected, except for a few cases.  Specifically, if no filter text has been
+            // provided, and this is not a preselect match then we will soft select it.  This
+            // happens when the completion list comes up implicitly and there is something in
+            // the MRU list.  In this case we do want to select it, but not with a hard
+            // selection.  Otherwise you can end up with the following problem:
+            //
+            //  dim i as integer =<space>
+            //
+            // Completion will comes up after = with 'integer' selected (Because of MRU).  We do
+            // not want 'space' to commit this.
+
+            if (ShouldSoftSelectItem(bestItem, filterText, triggerReason))
```

Comment:
this code changed as well.  i would have preferred we kept it as close to the original as possible. i.e. why not just return the boolean as before, then map that to this enum value.

Reply:
The Editor wants us to use NoChange as much as possible. We could use NoChange in most cases to make working most of unit tests but not all of them. To fix more cases, I had to restore more methods from the old code.

However, I think we should somehow move from Soft/Hard to Soft/Hard/NoChange values. I'm going to make more migrations to this direction in later PRs.

---
In reply to: [222088317](https://github.com/dotnet/roslyn/pull/29016#discussion_r222088317) [](ancestors = 222088317)